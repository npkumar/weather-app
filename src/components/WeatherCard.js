import React from "react";
import PropTypes from "prop-types";
import capitalize from "lodash/capitalize";
import dayjs from "dayjs";
import { makeStyles } from "@material-ui/core/styles";
import CardContent from "@material-ui/core/CardContent";
import Typography from "@material-ui/core/Typography";
import { Paper } from "@material-ui/core";
import {
  useApplication,
  KELVIN,
  CELSIUS,
  FAHRENHEIT,
} from "../services/Application";

const useStyles = makeStyles({
  root: {
    marginTop: 16,
    textAlign: "center",
  },
  title: {
    fontSize: 24,
    margin: 0,
  },
  smallTitle: {
    fontSize: 16,
  },
  description: {
    fontSize: 16,
    fontWeight: "bold",
  },
  smallDescription: {
    fontSize: 12,
  },
  dayOfWeek: {
    fontSize: 16,
  },
  smallDayOfWeek: {
    fontSize: 12,
  },
});

const WeatherCard = ({
  small,
  date,
  avgTempKelvin,
  avgTempFahrenheit,
  avgTempCelsius,
  ...otherProps
}) => {
  const classes = useStyles();
  const { application } = useApplication();

  const unitMap = {
    [KELVIN]: avgTempKelvin,
    [FAHRENHEIT]: avgTempFahrenheit,
    [CELSIUS]: avgTempCelsius,
  };

  return (
    <Paper className={classes.root} elevation={3} {...otherProps}>
      <CardContent>
        <Typography
          className={small ? classes.smallTitle : classes.title}
          gutterBottom
          data-testid="date"
        >
          {dayjs(date).format("MMM DD, YYYY")}
        </Typography>
        <Typography
          className={small ? classes.smallDayOfWeek : classes.dayOfWeek}
          gutterBottom
          data-testid="day"
        >
          {dayjs(date).format("dddd")}
        </Typography>
        <Typography
          component="p"
          color="secondary"
          className={small ? classes.smallDescription : classes.description}
          data-testid="unit"
        >
          {unitMap[application.unit]} {capitalize(application.unit)}
        </Typography>
      </CardContent>
    </Paper>
  );
};

WeatherCard.propTypes = {
  date: PropTypes.string.isRequired,
  avgTempKelvin: PropTypes.number.isRequired,
  avgTempFahrenheit: PropTypes.number.isRequired,
  avgTempCelsius: PropTypes.number.isRequired,
  small: PropTypes.bool,
};

WeatherCard.defaultProps = {
  small: false,
};

export default WeatherCard;
