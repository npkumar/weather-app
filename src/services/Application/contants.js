export const FAHRENHEIT = "fahrenheit";
export const KELVIN = "kelvin";
export const CELSIUS = "celsius";
