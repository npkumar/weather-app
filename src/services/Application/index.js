// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";

// Constants
export { FAHRENHEIT, CELSIUS, KELVIN } from "./contants";
