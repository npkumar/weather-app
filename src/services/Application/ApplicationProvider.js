import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";
import * as constants from "./contants";

// Actions
const SET_UNIT = "SET_UNIT";
const SET_INDEX = "SET_INDEX";

export const ApplicationContext = createContext(null);

const ApplicationProvider = ({ children }) => {
  const initialState = {
    unit: constants.FAHRENHEIT,
    index: 0,
  };

  function reducer(state, action) {
    switch (action.type) {
      case SET_UNIT: {
        return { ...state, unit: action.payload };
      }
      case SET_INDEX: {
        return { ...state, index: action.payload };
      }
      default:
        throw new Error();
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setUnit: (unit) => dispatch({ type: SET_UNIT, payload: unit }),
    setIndex: (index) => dispatch({ type: SET_INDEX, payload: index }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
