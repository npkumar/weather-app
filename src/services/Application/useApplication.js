import { useContext } from "react";
import { ApplicationContext } from "./ApplicationProvider";

const useApplication = () => useContext(ApplicationContext);

export default useApplication;
