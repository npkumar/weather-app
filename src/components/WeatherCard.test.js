import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ApplicationContext } from "../services/Application/ApplicationProvider";
import * as constants from "../services/Application/contants";
import WeatherCard from "./WeatherCard";

const value = {
  application: {
    unit: constants.FAHRENHEIT,
  },
};

const props = {
  date: "2020-07-23",
  avgTempKelvin: 300,
  avgTempFahrenheit: 200,
  avgTempCelsius: 100,
};

describe("WeatherCard", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <WeatherCard {...props} />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display date, day and temperature in Fahrenheit by default", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <WeatherCard {...props} />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("date")).toHaveTextContent("Jul 23, 2020");
    expect(getByTestId("day")).toHaveTextContent("Thursday");
    expect(getByTestId("unit")).toHaveTextContent("200 Fahrenheit");
  });

  it("should display date, day and temperature in Celsius if application unit is Celsius", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider
        value={{ application: { unit: constants.CELSIUS } }}
      >
        <WeatherCard {...props} />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("date")).toHaveTextContent("Jul 23, 2020");
    expect(getByTestId("day")).toHaveTextContent("Thursday");
    expect(getByTestId("unit")).toHaveTextContent("100 Celsius");
  });
});
