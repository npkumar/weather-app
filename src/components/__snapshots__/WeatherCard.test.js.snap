// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCard renders 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root makeStyles-root-1 MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root makeStyles-title-2 MuiTypography-body1 MuiTypography-gutterBottom"
        data-testid="date"
      >
        Jul 23, 2020
      </p>
      <p
        class="MuiTypography-root makeStyles-dayOfWeek-6 MuiTypography-body1 MuiTypography-gutterBottom"
        data-testid="day"
      >
        Thursday
      </p>
      <p
        class="MuiTypography-root makeStyles-description-4 MuiTypography-body1 MuiTypography-colorSecondary"
        data-testid="unit"
      >
        200 Fahrenheit
      </p>
    </div>
  </div>
</DocumentFragment>
`;
